modules/apps/redirect/redirect-service/src/main/java/com/liferay/redirect/internal/provider/RedirectProviderImpl.java

```diff
@@ -156,6 +160,52 @@ protected void setRedirectPatternEntries(
 		_redirectPatternEntries = redirectPatternEntries;
 	}
 
+	private String[] _getCrawlerUserAgents() {
+		return new String[0];
+	}
+
+	private boolean _isCrawlerUserAgent(String userAgent) {
+		for (String crawlerUserAgent : _getCrawlerUserAgents()) {
+			if (userAgent.contains(StringUtil.toLowerCase(crawlerUserAgent))) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private boolean _isPatternApplied(
+		RedirectPatternEntry redirectPatternEntry, String userAgent) {
+
+		if (Validator.isNull(redirectPatternEntry.getUserAgent()) ||
+			Validator.isNull(userAgent) ||
+			Objects.equals(
+				RedirectConstants.USER_AGENT_ALL,
+				redirectPatternEntry.getUserAgent())) {
+
+			return true;
+		}
+
+		if (_isCrawlerUserAgent(userAgent)) {
+			if (Objects.equals(
+					RedirectConstants.USER_AGENT_BOT,
+					redirectPatternEntry.getUserAgent())) {
+
+				return true;
+			}
+		}
+		else {
+			if (Objects.equals(
+					RedirectConstants.USER_AGENT_HUMAN,
+					redirectPatternEntry.getUserAgent())) {
+
+				return true;
+			}
+		}
+
+		return false;
+	}
+
```

Comment:
> [not a change request] By looking at the code, I wonder if we could simplify things by only having two kinds of patterns: "bot" and "any". As patterns are checked in a specific order, I think the "human" category is unnecessary, and it would simplify the matching logic a little bit. What do you think?

I do not see this point.
on the case that the redirection user agent is human and the user agent of the request is a bot, the redirection is not applied, 
I do not see how we can do this with "any" . We can discuss it on person if I'm missing something.

Reply:
As patterns are sorted, you can get the same result by putting the "bot" patterns first. If there's a bot specific pattern, it should have matched before getting to the "any" pattern.

By having 3 kinds of patterns there are weird situations, e.g.: I have two patterns, "human" and "any"; depending on the order of declaration in the configuration, the "human" pattern may never be used (if "any" comes before). So, in any case, matching order will change the behavior. To me, having 3 pattern types would make sense if we scanned the whole list of patterns and took the best match, but that's not how it works; it will simply take the first one that matches.

By having only 2 types of user agent, we are explicitly using declaration order as a matching criteria, which is simpler (and is already documented, so it is part of the "API").

Reply:
As we talked, we need 3 kinds of "user agents" because we do not know how the user will introduce the redirections

On the case if the user only insert redirections for bot 
On the case the user only insert redirections for humans (or not bots)
on the case for all
on a mix of them (we should be careful on the saving)

Reply:
I will simplify this code a little